src/encrypt.c

```diff
@@ -0,0 +1,306 @@
+#include "keys.h"
+
+#include <string.h>
+
+#include <openssl/ec.h>
+#include <openssl/evp.h>
+#include <openssl/rand.h>
+
+#include <openssl/err.h>
+
+// Writes an unsigned 32-bit integer in network byte order.
+static inline void
+ece_write_uint32_be(uint8_t* bytes, uint32_t value) {
+  bytes[0] = (value >> 24) & 0xff;
+  bytes[1] = (value >> 16) & 0xff;
+  bytes[2] = (value >> 8) & 0xff;
+  bytes[3] = value & 0xff;
+}
+
+// Encrypts a plaintext block with optional padding.
+static int
+ece_aes128gcm_encrypt_block(EVP_CIPHER_CTX* ctx, const ece_buf_t* key,
+                            const ece_buf_t* nonce, size_t counter,
+                            const ece_buf_t* block, const ece_buf_t* pad,
+                            ece_buf_t* record) {
+
+  int err = ECE_OK;
+
+  // Generate the IV for this record using the nonce.
+  uint8_t iv[ECE_NONCE_LENGTH];
+  ece_generate_iv(nonce->bytes, counter, iv);
+
+  // Encrypt the plaintext and padding.
+  if (EVP_EncryptInit_ex(ctx, EVP_aes_128_gcm(), NULL, key->bytes, iv) <= 0) {
+    err = ECE_ERROR_ENCRYPT;
+    goto end;
+  }
+  int blockLen = -1;
+  if (EVP_EncryptUpdate(ctx, record->bytes, &blockLen, block->bytes,
+                        block->length) <= 0 ||
+      blockLen != block->length) {
+    err = ECE_ERROR_ENCRYPT;
+    goto end;
+  }
+  int padLen = -1;
+  if (EVP_EncryptUpdate(ctx, &record->bytes[blockLen], &padLen, pad->bytes,
+                        pad->length) <= 0 ||
+      padLen != pad->length) {
+    err = ECE_ERROR_ENCRYPT;
+    goto end;
+  }
+  int finalLen = -1;
+  if (EVP_EncryptFinal_ex(ctx, &record->bytes[blockLen + padLen], &finalLen) <=
+        0 ||
+      finalLen) {
```

Comment:
Add a comment that `EncryptFinal` should not write any data, and check if it's OK to pass `NULL` for the second argument.